numpy/lib/_iotools.py

```diff
@@ -504,18 +504,23 @@ class StringConverter:
     """
     #
     _mapper = [(nx.bool_, str2bool, False),
-               (nx.integer, int, -1)]
+               (nx.int_, int, -1),]
 
     # On 32-bit systems, we need to make sure that we explicitly include
-    # nx.int64 since ns.integer is nx.int32.
-    if nx.dtype(nx.integer).itemsize < nx.dtype(nx.int64).itemsize:
+    # nx.int64 since ns.int_ is nx.int32.
+    if nx.dtype(nx.int_).itemsize < nx.dtype(nx.int64).itemsize:
         _mapper.append((nx.int64, int, -1))
 
-    _mapper.extend([(nx.floating, float, nx.nan),
-                    (nx.complexfloating, complex, nx.nan + 0j),
+    _mapper.extend([(nx.float64, float, nx.nan),
+                    (nx.complex128, complex, nx.nan + 0j),
                     (nx.longdouble, nx.longdouble, nx.nan),
                     (nx.unicode_, asunicode, '???'),
-                    (nx.string_, asbytes, '???')])
+                    (nx.string_, asbytes, '???'),
+                    # If a non-default dtype is passed, fall back to generic
+                    # ones (should only be used for the converter)
+                    (nx.integer, int, -1),
+                    (nx.floating, float, nx.nan),
+                    (nx.complexfloating, complex, nx.nan + 0j),])
```

Comment:
This is the most annoying part of the change. I think it is OK, basically... This only gets called when the dtype is was already fixed, in which case the dtype is not used, but only the converters is found.

Reply:
Turns out this is not ok:

This change has broken `genfromtxt` with dtype object:

Before:
```python
>>> np.genfromtxt(io.StringIO("1"), dtype=object)
array(b'1', dtype=object)
>>> np.genfromtxt(io.StringIO("what"), dtype=object)
array(b'what', dtype=object)
```

After:
```python
>>> np.genfromtxt(io.StringIO("1"), dtype=object)
array((1+0j), dtype=object)
>>> np.genfromtxt(io.StringIO("what"), dtype=object)
array(None, dtype=object)
```

Reply:
Ouch, subtle, I guess string being the last entry is important, probably switching the order will be OK, but have to look...

Reply:
Filed as #16189